File: Laravel/app/Http/Controllers/BdAuthController.php
<?php

namespace App\Http\Controllers;

use App\Bd\Auth;
use Illuminate\Http\Request;
use Illuminate\Http\Response;

/**
 * Mounts the whole tenant-auth flow at `/api/bd-auth/{action}`.
 *
 * `BD_AUTH_CALLBACK_URL` must point at this route's `callback` action and
 * be registered as a redirect URI on the BD site, or `auth/start` refuses.
 */
class BdAuthController extends Controller
{
    public function __invoke(Request $request, string $action)
    {
        $auth = Auth::fromConfig();
        if (! $auth) {
            return response('BD auth is not configured.', 503);
        }

        return match ($action) {
            'sign-in', 'sign-up' => $this->start($request, $auth, $action),
            'callback' => $this->callback($request, $auth),
            'sign-out' => $this->signOut($request, $auth),
            default => response('Not found', 404),
        };
    }

    private function start(Request $request, Auth $auth, string $intent)
    {
        $returnTo = $request->query('returnTo', '/my-account');

        try {
            return redirect()->away($auth->startUrl(
                $intent,
                is_string($returnTo) ? $returnTo : '/my-account',
                $request->query('loginHint'),
            ));
        } catch (\Throwable $e) {
            // Plain text, not an exception page: a failed redirect here is
            // almost always a missing `tenant_auth:public` scope or an
            // unregistered redirect URI, and the message says which.
            return response(
                "Sign-in could not start.\n\n".$e->getMessage()."\n\n".
                "Check that the API key carries the tenant_auth scope and that\n".
                'BD_AUTH_CALLBACK_URL is registered as a redirect URI.',
                502
            )->header('Content-Type', 'text/plain; charset=utf-8');
        }
    }

    private function callback(Request $request, Auth $auth)
    {
        $code = $request->query('code');
        $state = $request->query('state');

        if (! is_string($code) || $code === '') {
            return response('Missing code', 400);
        }
        if (! is_string($state) || trim($state) === '') {
            return response('Missing OAuth state', 400);
        }

        try {
            $session = $auth->exchange($code, $state);
        } catch (\Throwable $e) {
            return response('Sign-in failed: '.$e->getMessage(), 400);
        }

        $expiresAt = strtotime($session['expiresAt']) ?: (time() + 60 * 60 * 24 * 7);
        $minutes = max(1, (int) round(($expiresAt - time()) / 60));

        return redirect($auth->returnToFromState($state, '/my-account'))
            ->withCookie(cookie(
                name: config('bd.session_cookie'),
                value: $session['sessionToken'],
                minutes: $minutes,
                path: '/',
                domain: null,
                secure: $request->isSecure(),
                httpOnly: true,
                raw: false,
                sameSite: 'lax',
            ));
    }

    private function signOut(Request $request, Auth $auth): Response
    {
        $auth->signOut();

        $response = $request->isMethod('POST')
            ? response()->json(['ok' => true])
            : redirect('/');

        return $response->withCookie(
            cookie()->forget(config('bd.session_cookie'), '/')
        );
    }
}

File: Phoenix/lib/bd_starter_web/controllers/bd_auth_controller.ex
defmodule BdStarterWeb.BdAuthController do
  use BdStarterWeb, :controller

  alias BdStarter.Bd
  alias BdStarter.Bd.Auth

  @doc """
  The whole tenant-auth flow, mounted at `/api/bd-auth/:action`.

  `BD_AUTH_CALLBACK_URL` must point at this route's `callback` action AND be
  registered as a redirect URI on the BD site, or `auth/start` refuses.
  """
  def handle(conn, %{"action" => action} = params) do
    case {Bd.client(), action} do
      {nil, _} ->
        send_resp(conn, 503, "BD auth is not configured.")

      {client, a} when a in ["sign-in", "sign-up"] ->
        start(conn, client, a, params)

      {client, "callback"} ->
        callback(conn, client, params)

      {client, "sign-out"} ->
        sign_out(conn, client)

      _ ->
        send_resp(conn, 404, "Not found")
    end
  end

  defp start(conn, client, intent, params) do
    return_to = Map.get(params, "returnTo", "/my-account")

    case Auth.start_url(client, intent, return_to, Map.get(params, "loginHint")) do
      {:ok, url} ->
        redirect(conn, external: url)

      {:error, reason} ->
        # Plain text, not an exception page: this almost always means the API
        # key is missing the `tenant_auth:public` scope or the callback URL
        # isn't registered, and the message should say so.
        conn
        |> put_resp_content_type("text/plain")
        |> send_resp(502, """
        Sign-in could not start.

        #{inspect(reason)}

        Check that the API key carries the tenant_auth scope and that
        BD_AUTH_CALLBACK_URL is registered as a redirect URI on the site.
        """)
    end
  end

  defp callback(conn, client, params) do
    with code when is_binary(code) and code != "" <- Map.get(params, "code"),
         state when is_binary(state) and state != "" <- Map.get(params, "state"),
         {:ok, %{"sessionToken" => token} = session} <- Auth.exchange(client, code, state) do
      conn
      |> put_resp_cookie(cookie_name(), token,
        http_only: true,
        same_site: "Lax",
        secure: conn.scheme == :https,
        max_age: max_age(session["expiresAt"])
      )
      |> redirect(to: Auth.return_to_from_state(state, "/my-account"))
    else
      nil -> send_resp(conn, 400, "Missing code or OAuth state")
      "" -> send_resp(conn, 400, "Missing code or OAuth state")
      {:error, reason} -> send_resp(conn, 400, "Sign-in failed: #{inspect(reason)}")
      _ -> send_resp(conn, 400, "Sign-in failed")
    end
  end

  defp sign_out(conn, client) do
    Auth.sign_out(client)

    conn = delete_resp_cookie(conn, cookie_name(), path: "/")

    if conn.method == "POST" do
      json(conn, %{ok: true})
    else
      redirect(conn, to: ~p"/")
    end
  end

  defp max_age(expires_at) when is_binary(expires_at) do
    case DateTime.from_iso8601(expires_at) do
      {:ok, dt, _} -> max(60, DateTime.diff(dt, DateTime.utc_now()))
      _ -> default_max_age()
    end
  end

  defp max_age(_), do: default_max_age()
  defp default_max_age, do: 60 * 60 * 24 * 7

  defp cookie_name, do: Application.get_env(:bd_starter, :bd_session_cookie, "bd_session")
end
